## Module 1 Homework

In this homework, we're going to download finance data from various sources and make simple calculations or analysis.

### Question 1: [Index] S&P 500 Stocks Added to the Index

**Which year had the highest number of additions?**

Using the list of S&P 500 companies from Wikipedia's [S&P 500 companies page](https://en.wikipedia.org/wiki/List_of_S%26P_500_companies), download the data including the year each company was added to the index.

Hint: you can use [pandas.read_html](https://pandas.pydata.org/docs/reference/api/pandas.read_html.html) to scrape the data into a DataFrame.

Steps:
1. Create a DataFrame with company tickers, names, and the year they were added.
2. Extract the year from the addition date and calculate the number of stocks added each year.
3. Which year had the highest number of additions (1957 doesn't count, as it was the year when the S&P 500 index was founded)? Write down this year as your answer (the most recent one, if you have several records).

*Context*: 
> "Following the announcement, all four new entrants saw their stock prices rise in extended trading on Friday" - recent examples of S&P 500 additions include DASH, WSM, EXE, TKO in 2025 ([Nasdaq article](https://www.nasdaq.com/articles/sp-500-reshuffle-dash-tko-expe-wsm-join-worth-buying)).

In [1]:
import pandas as pd
from datetime import datetime

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_S%26P_500_companies'
tables = pd.read_html(url)
earnings_data = tables[0]

In [3]:
earnings_data

Symbol             Security             GICS Sector  \
0      MMM                   3M             Industrials   
1      AOS          A. O. Smith             Industrials   
2      ABT  Abbott Laboratories             Health Care   
3     ABBV               AbbVie             Health Care   
4      ACN            Accenture  Information Technology   
..     ...                  ...                     ...   
498    XYL           Xylem Inc.             Industrials   
499    YUM          Yum! Brands  Consumer Discretionary   
500   ZBRA   Zebra Technologies  Information Technology   
501    ZBH        Zimmer Biomet             Health Care   
502    ZTS               Zoetis             Health Care   

                                GICS Sub-Industry    Headquarters Location  \
0                        Industrial Conglomerates    Saint Paul, Minnesota   
1                               Building Products     Milwaukee, Wisconsin   
2                           Health Care Equipment  North Chicago, Illinois   
3                                   Biotechnology  North Chicago, Illinois   
4                  IT Consulting & Other Services          Dublin, Ireland   
..                                            ...                      ...   
498  Industrial Machinery & Supplies & Components   White Plains, New York   
499                                   Restaurants     Louisville, Kentucky   
500            Electronic Equipment & Instruments   Lincolnshire, Illinois   
501                         Health Care Equipment          Warsaw, Indiana   
502                               Pharmaceuticals   Parsippany, New Jersey   

     Date added      CIK      Founded  
0    1957-03-04    66740         1902  
1    2017-07-26    91142         1916  
2    1957-03-04     1800         1888  
3    2012-12-31  1551152  2013 (1888)  
4    2011-07-06  1467373         1989  
..          ...      ...          ...  
498  2011-11-01  1524472         2011  
499  1997-10-06  1041061         1997  
500  2019-12-23   877212         1969  
501  2001-08-07  1136869         1927  
502  2013-06-21  1555280         1952  

[503 rows x 8 columns]

In [4]:
(earnings_data
 .rename(columns=lambda x: x.replace(" ", "_"))
 .assign(Date_added=lambda x: pd.to_datetime(x['Date_added']))
 .assign(Year_added=lambda x: x['Date_added'].dt.year)
 .query("Year_added!=1957")
 .groupby('Year_added')
 .agg(Count=('Symbol','count'))
 .reset_index()
 .sort_values(by='Count', ascending=False)
 .head(1)
)

Year_added  Count
47        2017     23

**Answer**: 2017

*Additional*: How many current S&P 500 stocks have been in the index for more than 20 years? When stocks are added to the S&P 500, they usually experience a price bump as investors and index funds buy shares following the announcement.

In [5]:
(earnings_data
 .rename(columns=lambda x: x.replace(" ", "_"))
 .assign(Date_added=lambda x: pd.to_datetime(x['Date_added']))
 .assign(Years_in_list=lambda x: (datetime.now() - x['Date_added']).dt.days / 365.25)
 .query("Years_in_list>=20") 
).shape[0]

219

There are 219 stocks in S&P500 that have been in the index for more than 20 years.

### Question 2. [Macro] Indexes YTD (as of 1 May 2025)

**How many indexes (out of 10) have better year-to-date returns than the US (S&P 500) as of May 1, 2025?**

Using Yahoo Finance World Indices data, compare the year-to-date (YTD) performance (1 January-1 May 2025) of major stock market indexes for the following countries:
* United States - S&P 500 (^GSPC)
* China - Shanghai Composite (000001.SS)
* Hong Kong - HANG SENG INDEX (^HSI)	
* Australia - S&P/ASX 200 (^AXJO)
* India - Nifty 50 (^NSEI)
* Canada - S&P/TSX Composite (^GSPTSE)
* Germany - DAX (^GDAXI)
* United Kingdom - FTSE 100 (^FTSE)
* Japan - Nikkei 225 (^N225)
* Mexico - IPC Mexico (^MXX)
* Brazil - Ibovespa (^BVSP)

*Hint*: use start_date='2025-01-01' and end_date='2025-05-01' when downloading daily data in yfinance

Context: 
> [Global Valuations: Who's Cheap, Who's Not?](https://simplywall.st/article/beyond-the-us-global-markets-after-yet-another-tariff-update) article suggests "Other regions may be growing faster than the US and you need to diversify."

Reference: Yahoo Finance World Indices - https://finance.yahoo.com/world-indices/

In [6]:
import yfinance as yf

index_tickers = ["^GSPC", "000001.SS", "^HSI", "^AXJO", "^NSEI", "^GSPTSE", "^GDAXI", "^FTSE", "^N225", "^MXX", "^BVSP"]
index_data = yf.download(index_tickers, start="2025-01-01", end="2025-05-01")
index_data

YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  11 of 11 completed


Price             Close                                                    \
Ticker        000001.SS        ^AXJO     ^BVSP        ^FTSE        ^GDAXI   
Date                                                                        
2025-01-01          NaN          NaN       NaN          NaN           NaN   
2025-01-02  3262.561035  8201.200195  120125.0  8260.099609  20024.660156   
2025-01-03  3211.429932  8250.500000  118533.0  8224.000000  19906.080078   
2025-01-06  3206.923096  8288.500000  120022.0  8249.700195  20216.189453   
2025-01-07  3229.644043  8285.099609  121163.0  8245.299805  20340.570312   
...                 ...          ...       ...          ...           ...   
2025-04-24  3297.288086  7968.200195  134580.0  8407.400391  22064.509766   
2025-04-25  3295.060059          NaN  134739.0  8415.299805  22242.449219   
2025-04-28  3288.415039  7997.100098  135016.0  8417.299805  22271.669922   
2025-04-29  3286.655029  8070.600098  135093.0  8463.500000  22425.830078   
2025-04-30  3279.031006  8126.200195  135067.0  8494.900391  22496.980469   

Price                                                              \
Ticker            ^GSPC       ^GSPTSE          ^HSI          ^MXX   
Date                                                                
2025-01-01          NaN           NaN           NaN           NaN   
2025-01-02  5868.549805  24898.000000  19623.320312  49765.199219   
2025-01-03  5942.470215  25073.500000  19760.269531  48957.238281   
2025-01-06  5975.379883  24999.800781  19688.289062  49493.558594   
2025-01-07  5909.029785  24929.900391  19447.580078  50085.500000   
...                 ...           ...           ...           ...   
2025-04-24  5484.770020  24727.500000  21909.759766  56382.000000   
2025-04-25  5525.209961  24710.500000  21980.740234  56720.121094   
2025-04-28  5528.750000  24798.599609  21971.960938  56980.128906   
2025-04-29  5560.830078  24874.500000  22008.109375  55613.429688   
2025-04-30  5569.060059  24841.699219  22119.410156  56259.281250   

Price                     ...    Volume                                        \
Ticker             ^N225  ...     ^AXJO       ^BVSP         ^FTSE      ^GDAXI   
Date                      ...                                                   
2025-01-01           NaN  ...       NaN         NaN           NaN         NaN   
2025-01-02           NaN  ...  304400.0   9373600.0  4.222199e+08  52445600.0   
2025-01-03           NaN  ...  329100.0   9804400.0  7.425039e+08  44372900.0   
2025-01-06  39307.050781  ...   52200.0   9685600.0  7.662447e+08  70784900.0   
2025-01-07  40083.300781  ...  424300.0  11116400.0  7.415068e+08  62020000.0   
...                  ...  ...       ...         ...           ...         ...   
2025-04-24  35039.148438  ...  639100.0  14113400.0  1.126606e+09  62636800.0   
2025-04-25  35705.738281  ...       NaN  13051800.0  8.027340e+08  70917400.0   
2025-04-28  35839.988281  ...  769000.0  11449700.0  7.387417e+08  55883200.0   
2025-04-29           NaN  ...  710800.0  12761100.0  6.559248e+08  75547100.0   
2025-04-30  36045.378906  ...  889500.0  15452400.0  1.589592e+09  99189800.0   

Price                                                                          \
Ticker             ^GSPC      ^GSPTSE          ^HSI         ^MXX        ^N225   
Date                                                                            
2025-01-01           NaN          NaN           NaN          NaN          NaN   
2025-01-02  3.621680e+09  215089400.0  4.033400e+09   87535300.0          NaN   
2025-01-03  3.667340e+09  186569100.0  3.393800e+09  112782300.0          NaN   
2025-01-06  4.940120e+09  239976800.0  2.465700e+09  139872100.0  137900000.0   
2025-01-07  4.517330e+09  237759800.0  3.581000e+09  142173400.0  127000000.0   
...                  ...          ...           ...          ...          ...   
2025-04-24  4.697710e+09  224419200.0  2.985800e+09  249950000.0  137100000.0   
2025-04-2

In [7]:
(index_data
 .filter(like="Close")
 .bfill()
 .iloc[[0,-1]]
 .T
 .reset_index() 
 .pipe(lambda x: x.set_axis([*x.columns[:-2], 'Start', 'End'], axis=1))
 .assign(Growth=lambda x: ((x['End'] / x['Start']) - 1) * 100)
 .sort_values('Growth', ascending=False)
 [['Ticker', 'Growth']]
)

Ticker     Growth
8        ^MXX  13.049444
7        ^HSI  12.720018
2       ^BVSP  12.438710
4      ^GDAXI  12.346378
3       ^FTSE   2.842590
10      ^NSEI   2.490424
0   000001.SS   0.504817
6     ^GSPTSE  -0.226126
1       ^AXJO  -0.914500
5       ^GSPC  -5.103301
9       ^N225  -8.297931

**Answer**: 9

### Question 3. [Index] S&P 500 Market Corrections Analysis


**Calculate the median duration (in days) of significant market corrections in the S&P 500 index.**

For this task, define a correction as an event when a stock index goes down by **more than 5%** from the closest all-time high maximum.

Steps:
1. Download S&P 500 historical data (1950-present) using yfinance
2. Identify all-time high points (where price exceeds all previous prices)
3. For each pair of consecutive all-time highs, find the minimum price in between
4. Calculate drawdown percentages: (high - low) / high × 100
5. Filter for corrections with at least 5% drawdown
6. Calculate the duration in days for each correction period
7. Determine the 25th, 50th (median), and 75th percentiles for correction durations

*Context:* 
> * Investors often wonder about the typical length of market corrections when deciding "when to buy the dip" ([Reddit discussion](https://www.reddit.com/r/investing/comments/1jrqnte/when_are_you_buying_the_dip/?rdt=64135)).

> * [A Wealth of Common Sense - How Often Should You Expect a Stock Market Correction?](https://awealthofcommonsense.com/2022/01/how-often-should-you-expect-a-stock-market-correction/)

*Hint (use this data to compare with your results)*: Here is the list of top 10 largest corrections by drawdown:
* 2007-10-09 to 2009-03-09: 56.8% drawdown over 517 days
* 2000-03-24 to 2002-10-09: 49.1% drawdown over 929 days
* 1973-01-11 to 1974-10-03: 48.2% drawdown over 630 days
* 1968-11-29 to 1970-05-26: 36.1% drawdown over 543 days
* 2020-02-19 to 2020-03-23: 33.9% drawdown over 33 days
* 1987-08-25 to 1987-12-04: 33.5% drawdown over 101 days
* 1961-12-12 to 1962-06-26: 28.0% drawdown over 196 days
* 1980-11-28 to 1982-08-12: 27.1% drawdown over 622 days
* 2022-01-03 to 2022-10-12: 25.4% drawdown over 282 days
* 1966-02-09 to 1966-10-07: 22.2% drawdown over 240 days

In [8]:
sp500 = yf.download('^GSPC', start='1950-01-01')['Close']
all_time_highs = sp500.cummax()
sp500['Is_ATH'] = (sp500 == all_time_highs)
ath_dates = sp500.index[sp500['Is_ATH']].tolist()

[*********************100%***********************]  1 of 1 completed


In [19]:
corrections = []

for i in range(len(ath_dates)-1):
    start = ath_dates[i]
    end = ath_dates[i+1]
    high = all_time_highs.loc[ath_dates[i]].iloc[0]
    low = sp500.loc[start:end].min().iloc[0]
    
    # Calculate drawdown
    drawdown = (high - low)/high * 100
    if isinstance(drawdown, pd.Series):
        drawdown = drawdown.iloc[0]  # Handle series conversion
    
    if drawdown >= 5:
        min_date = sp500.loc[start:end].idxmin().iloc[0]
        duration = (min_date - start).days
        corrections.append({
            'Start Date': start.strftime('%Y-%m-%d'),
            'End Date': min_date.strftime('%Y-%m-%d'),
            'Duration (Days)': duration,
            'Drawdown %': round(drawdown, 1)
        })

# Create formatted dataframe
corrections_df = pd.DataFrame(corrections)
print(corrections_df.head(10))
corrections_df.sort_values(by='Drawdown %', ascending=False)

   Start Date    End Date  Duration (Days)  Drawdown %
0  1950-06-12  1950-07-17               35        14.0
1  1950-11-24  1950-12-04               10         6.5
2  1951-05-03  1951-06-29               57         8.1
3  1951-10-15  1951-11-23               39         6.1
4  1952-01-22  1952-02-20               29         6.4
5  1952-08-08  1952-10-22               75         6.8
6  1953-01-05  1953-09-14              252        14.8
7  1955-01-03  1955-01-17               14         5.9
8  1955-03-04  1955-03-14               10         6.8
9  1955-09-23  1955-10-11               18        10.6


Start Date    End Date  Duration (Days)  Drawdown %
56  2007-10-09  2009-03-09              517        56.8
54  2000-03-24  2002-10-09              929        49.1
24  1973-01-11  1974-10-03              630        48.2
22  1968-11-29  1970-05-26              543        36.1
65  2020-02-19  2020-03-23               33        33.9
..         ...         ...              ...         ...
25  1980-09-22  1980-09-29                7         5.3
67  2021-09-02  2021-10-04               32         5.2
23  1972-08-14  1972-10-16               63         5.1
50  1999-04-12  1999-04-19                7         5.1
39  1991-08-28  1991-10-09               42         5.0

[71 rows x 4 columns]

In [20]:
print(corrections_df['Duration (Days)'].describe(percentiles=[.25, .5, .75]))

count     71.000000
mean     113.098592
std      179.073341
min        7.000000
25%       21.500000
50%       39.000000
75%       89.000000
max      929.000000
Name: Duration (Days), dtype: float64


**Answer**: 39

### Question 4.  [Stocks] Earnings Surprise Analysis for Amazon (AMZN)


**Calculate the median 2-day percentage change in stock prices following positive earnings surprises days.**

Steps:
1. Load earnings data from CSV ([ha1_Amazon.csv](ha1_Amazon.csv)) containing earnings dates, EPS estimates, and actual EPS. Make sure you are using the correct delimiter to read the data, such as in this command ```python pandas.read_csv("ha1_Amazon.csv", delimiter=';') ```
2. Download complete historical price data using yfinance
3. Calculate 2-day percentage changes for all historical dates: for each sequence of 3 consecutive trading days (Day 1, Day 2, Day 3), compute the *return* as Close_Day3 / Close_Day1 - 1. (Assume Day 2 may correspond to the earnings announcement.)
4. Identify positive earnings surprises (where "actual EPS > estimated EPS" OR "Surprise (%)>0")
5. Calculate 2-day percentage changes following positive earnings surprises. Show your answer in % (closest number to the 2nd digit): *return* * 100.0
6. (Optional) Compare the median 2-day percentage change for positive surprises vs. all historical dates. Do you see the difference?

Context: Earnings announcements, especially when they exceed analyst expectations, can significantly impact stock prices in the short term.

Reference: Yahoo Finance earnings calendar - https://finance.yahoo.com/calendar/earnings?symbol=AMZN


In [29]:
# Load and clean earnings data
earnings = pd.read_csv('data/ha1_Amazon.csv', sep=';').iloc[:-1, :]  # Exclude the last row which is a summary row
earnings['Earnings Date'] = pd.to_datetime(earnings['Earnings Date'].str.split(' at').str[0], errors='coerce')

# Clean numeric columns
numeric_cols = ['EPS Estimate', 'Reported EPS', 'Surprise (%)']
for col in numeric_cols:
    earnings[col] = pd.to_numeric(earnings[col].str.replace('[^-.0-9]', '', regex=True), errors='coerce')

# Download historical prices
amzn = yf.download('AMZN', start='1997-05-15')['Close'].reset_index()
amzn.columns = ['Date', 'Price']

# Calculate 2-day returns for ALL dates
amzn['2_day_pct_change'] =((amzn['Price'].shift(-2) / amzn['Price']) - 1)

positive_surprises = earnings[
    (earnings['Reported EPS'] > earnings['EPS Estimate']) | 
    (earnings['Surprise (%)'] > 0)
].copy()
#

dates_to_consider = positive_surprises['Earnings Date'].tolist()
# Filter AMZN data for the dates of interest
amzn_filtered = amzn[amzn['Date'].isin(dates_to_consider)].copy()

#( amzn_filtered['Price'].shift(-2)/amzn_filtered['Price'] -1).median()
amzn_filtered['2_day_pct_change'].median()*100

[*********************100%***********************]  1 of 1 completed


np.float64(0.2672266474036067)

In [30]:
# Merge earnings dates with price data
merged = pd.merge_asof(
    earnings.sort_values('Earnings Date'),
    amzn.sort_values('Date'),
    left_on='Earnings Date',
    right_on='Date',
    direction='forward'
)

# Filter positive earnings surprises
positive_surprises = merged[
    ((merged['Reported EPS'] > merged['EPS Estimate']) | (merged['Surprise (%)'] > 0)) &
    merged['2_day_pct_change'].notna()
].copy()

# Calculate median 2-day percentage changes
median_positive = positive_surprises['2_day_pct_change'].median()
median_all = amzn['2_day_pct_change'].median()

median_positive, median_all

(np.float64(0.002672266474036067), np.float64(0.0016375198550677705))

In [31]:
merged = pd.merge_asof(
    earnings.sort_values('Earnings Date'),
    amzn.sort_values('Date'),
    left_on='Earnings Date',
    right_on='Date',
    direction='forward',
    tolerance=pd.Timedelta('2D')
)
positive_surprises = merged[
    ((merged['Reported EPS'] > merged['EPS Estimate']) | 
    (merged['Surprise (%)'] > 0)) &
    merged['2_day_pct_change'].notna()
]

# Calculate comparison metrics
all_days_median = amzn['2_day_pct_change'].median()*100
positive_surprises_median = positive_surprises['2_day_pct_change'].median()*100

print(f"Median 2-day return after positive surprises: {positive_surprises_median:.2%}")
print(f"Median 2-day return for all trading days: {all_days_median:.2%}")

# Optional: Display key examples
print("\nNotable examples:")
print(positive_surprises[['Earnings Date', 'Surprise (%)', '2_day_pct_change']].head(10))

Median 2-day return after positive surprises: 26.72%
Median 2-day return for all trading days: 16.38%

Notable examples:
  Earnings Date  Surprise (%)  2_day_pct_change
0    1997-07-10         13.33         -0.166665
1    1997-10-27         13.29          0.262192
2    1998-01-22         11.41         -0.062881
3    1998-04-27         13.92          0.154078
4    1998-07-22          1.34         -0.072761
5    1998-10-28         15.53          0.080085
6    1999-01-26         22.22          0.067608
7    1999-04-28         19.92         -0.110788
8    1999-07-21          0.47         -0.086697
9    1999-10-27          7.54         -0.069959
